ast/check.go

```diff
@@ -685,8 +685,7 @@ func (rc *refChecker) checkRef(curr *TypeEnv, node *typeTreeNode, ref Ref, idx i
 	// Handle dynamic ref operands.
 	switch value := head.Value.(type) {
 
-	case Var:
-
+	case Var, Number, Boolean:
```

Comment:
I'm not sure these extra cases are doing anything.

If the value is number or boolean then exist != nil and the comparison between tpe and exist should always be false because the keys of nodes in the type tree are always strings. Or is it possible for there to be non-leaf nodes in the type tree w/ keys that are number/boolean? If that's the case, what about other types (null/array/object/set)?

Reply:
I think this was because previous partial-object rules like
```
p[4] { true }
```
now are parsed as single-value rules with head ref `p[4]`. I think what I saw in the tests was stuff like
```
q := p[4][_]
```
and when that ref is checked, the `4` is _not_ the leaf node. But I will have another, closer look at this.

Also yes, at some point, I had been taking care of other types; but dialled that back. But if this ☝️ holds, we should take care of them. 🤔

Reply:
You're referring to the type tree, not how refs are checked. I'll circle in on this, to figure out what's what. I suspect L661 is the condition to adapt, not this one.

Reply:
I was assuming the type tree would not be affected but that was not true. For example:

```rego
package x

p[7] { true }

q[7] = 8 { true }

r = {7: 8}
```

**Before**:

```
$ go run main.go check ~/x.rego
 <nil> <nil>
  data <nil>
   "x" <nil>
    "q" object[number: number]
    "r" object<7: number>
    "p" set[number]
```

**After**:

(Ignore the difference in order)

```
$ go run main.go check ~/x.rego
 <nil> <nil>
  data <nil>
   "x" <nil>
    "p" set[number]
    "q" <nil>
     7 number
    "r" object<7: number>
```

Of course this makes sense now that we just have single-value and multi-value rules. Here's a slightly more complete example of these edge cases:

```
package x

p[7] = 8 { true }
p[false] = 8 { true }
p[null] = 8 { true }

p1[x] = 8 { x = 7 }
p1[x] = 8 { x = false }
p1[x] = 8 { x = null }

p2[[7]] = 8 { true }
p2[{7: 8}] = 9 { true }

p3[x] = 8 { x = [7] }
p3[x] = 9 { x = {7: 8} }

p4[[x]] = 8 { x = 7 }
p4[{x: y}] = 9 { x = 7; y = 8 }
```

```
$ go run main.go check ~/x.rego
 <nil> <nil>
  data <nil>
   "x" <nil>
    "p" <nil>
     7 number
     null number
     false number
    "p2" <nil>
     [7] number
     {7: 8} number
    "p3" object[any<array<number>, object<7: number>>: number]
    "p4" object[any<array<number>, object[number: number]>: number]
    "p1" object[any<null, boolean, number>: number]
```

Prefixing all of these rules w/ a string prefix produces the expected results and evaluation seems to do the right thing.